src/condition/ssrf_detector.cpp

```diff
@@ -0,0 +1,279 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include "condition/ssrf_detector.hpp"
+#include "exception.hpp"
+#include "iterator.hpp"
+#include "uri_utils.hpp"
+#include "utils.hpp"
+
+using namespace std::literals;
+
+namespace ddwaf {
+
+namespace {
+
+constexpr std::array<std::string_view, 10> dangerous_ips{"169.254.0.0/16", "127.0.0.1/32",
+    "10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16", "100.64.0.0/10", "::1/128", "fc00::/7",
+    "fe80::/10", "2001:db8:1234:1a00::/56"};
+
+constexpr const std::array<std::string_view, 10> dangerous_domains{
+    "metadata.google", "burpcollaborator.net", ".local", ".internal", "ram.aliyuncs.com",
+    "ifconfig.pro", "localhost", "fuf.me", "localtest.me", "ulh.us" // Legacy domains
+};
+
+constexpr const std::array<std::string_view, 4> authorised_schemes{"https", "http", "ftps", "ftp"};
+
+constexpr const auto &npos = std::string_view::npos;
+
+using ssrf_result = std::optional<std::pair<std::string, std::vector<std::string>>>;
+
+bool detect_parameter_injection(
+    const uri_decomposed &uri, std::string_view param, std::size_t param_index)
+{
+    const auto path_end = uri.path_index + uri.path.size();
+
+    // Check if the application is giving full control of the path to the user
+    // either fully or after a forward-slash.
+    if (!uri.path.empty()) {
+        //  scheme://userinfo@host:port/path?query#fragment
+        //                             ─────
+        if (uri.path.size() == param.size()) {
+            return false;
+        }
+        //  scheme://userinfo@host:port/path?query#fragment
+        //                              ────
+        if (uri.path[0] == '/' && param_index == uri.path_index + 1 &&
+            param.size() == uri.path.size() - 1) {
+            return false;
+        }
+    }
+
+    // REST parameter injection will involve introducing a / to the URL, however, such
+    // slashes are allowed after the ? or the #. Check if the slash is within the path
+    //
+    //  scheme://userinfo@host:port/path?query#fragment
+    //                             ─────
+    if (auto slash_index = param.find('/'); slash_index != npos) {
+        slash_index += param_index;
+        if (slash_index < path_end) {
+            // The path is partially under control of the user, this might be intentional
+            // so lets check for a possible LFI
+            auto relative_dir_index = param.find("..");
+            while (relative_dir_index != npos) {
+
+                // We found '..', check if it's enclosed by '/'
+                auto dir_index = relative_dir_index + param_index;
+                if (dir_index < path_end && (dir_index + 2) < uri.raw.size() &&
+                    uri.raw[dir_index - 1] == '/' && uri.raw[dir_index + 2] == '/') {
```

Comment:
isn't \ a possibility on windows? what about %2F (and %5C for that matter), since you're comparing against the raw (undecoded) uri?

Reply:
According to the RFC the back-slash isn't allowed as a path separator, so I omitted that entirely. The use of  %2F is a bit nuanced as in theory it's not meant to be used as a path separator but rather as a way to introduce a forward-slash that doesn't clash with the reserved forward-slash. In practice some parsers might actually consider %2F as a path separator. What do you think @Taiki-San ?

Reply:
In Java, for instance, you have:

```
groovy:000> new URI('https://example.com/a%2fb').path
===> /a/b
```

you'd have to use rawPath do get the undecoded version, so you can slip a path separator this way

Reply:
As discussed before, it might make sense to simply flag and path injection containing `%2f`... cc: @Taiki-San

Reply:
I added a check to flag a parameter overlapping the path and containing %2f as an injection.